pkgs/games/ldmud/default.nix

```diff
@@ -0,0 +1,75 @@
+{ lib, fetchFromGitHub, stdenv, autoreconfHook, pkg-config, bison, libiconv
+, pcre, libgcrypt, ipv6Support ? false, mccpSupport ? false, zlib
+, mysqlSupport ? false, libmysqlclient, postgresSupport ? false, postgresql
+, sqliteSupport ? false, sqlite, json_c, libxml2, tlsSupport ? false, openssl
+, pythonSupport ? false, python310 }:
+
+stdenv.mkDerivation rec {
+  pname = "ldmud";
+  version = "3.6.6";
+
+  src = fetchFromGitHub {
+    owner = pname;
+    repo = pname;
+    rev = version;
+    sha256 = "sha256-2TaFt+T9B5Df6KWRQcbhY1E1D6NISb0oqLgyX47f5lI=";
+  };
+
+  sourceRoot = "source/src";
+
+  nativeBuildInputs = [ autoreconfHook pkg-config bison libgcrypt pcre json_c libxml2 ]
+    ++ lib.optional mccpSupport zlib
+    ++ lib.optional mysqlSupport libmysqlclient
+    ++ lib.optional postgresSupport postgresql
+    ++ lib.optional sqliteSupport sqlite
+    ++ lib.optional tlsSupport openssl
+    ++ lib.optional pythonSupport python310;
+  buildInputs = lib.optionals stdenv.isDarwin [ libiconv ];
+
+  # To support systems without autoconf LD puts its configure.ac in a non-default
+  # location and uses a helper script. We skip that script and symlink the .ac
+  # file to where the autoreconfHook find it.
+  preAutoreconf = ''
+    ln -fs ./autoconf/configure.ac ./
+  '';
+
+  configureFlags = [
+    "--enable-erq=xerq"
+    "--enable-filename-spaces"
+    "--enable-use-json"
+    "--enable-use-xml=xml2"
+    (lib.enableFeature ipv6Support "use-ipv6")
+    (lib.enableFeature mccpSupport "use-mccp")
+    (lib.enableFeature mysqlSupport "use-mysql")
+    (lib.enableFeature postgresSupport "use-pgsql")
+    (lib.enableFeature sqliteSupport "use-sqlite")
+    (lib.enableFeatureAs tlsSupport "use-tls" "ssl")
+    (lib.enableFeature pythonSupport "use-python")
+  ];
+
+  preConfigure = lib.optionalString mysqlSupport ''
+    export CPPFLAGS="-I${lib.getDev libmysqlclient}/include/mysql"
+    export LDFLAGS="-L${libmysqlclient}/lib/mysql"
+  '';
+
+  installTargets = [ "install-driver" "install-utils" "install-headers" ];
+
+  meta = with lib; {
+    description = "LDMud is a gamedriver for LPMuds. (LPC compiler, interpreter and runtime environment.)";
+    homepage = "https://ldmud.eu";
+    changelog = "https://github.com/ldmud/ldmud/blob/${version}/HISTORY";
+    longDescription = ''
+      LDMud started as a project to clean up and modernize Amylaar's LPMud
+      gamedriver. Primary goals are full documentation, a commented source body
+      and out-of-the-box support for the major mudlibs, of which the commented
+      source body has been pretty much completed. During the course of work
+      a lot of bug fixes and improvements found their way into the driver - much
+      more than originally expected, and definitely enough to make LDMud
+      a driver in its own right.
+    '';
+    # See https://github.com/ldmud/ldmud/blob/master/COPYRIGHT
+    license = licenses.free;
```

Comment:
At the risk of starting "one of those debates". `free` is for "Unspecified free software license". I expect that's following the FSF/OSI/etc definition where "noncommercial" (https://github.com/ldmud/ldmud/blob/64b3588a13fb0c761da62d6deb56dfa380b03c6f/COPYRIGHT#L13-L14) licenses are not "free" (https://en.wikipedia.org/wiki/Free-software_license)

Should this be its own license instead?

Reply:
@viraptor This is definitely outside my wheelhouse and I could use some advice. I think there are three options:

* Keep it as is
  * _Upsides:_ the upstream license allows redistribution which seems most important for the nixpkgs context, and we don't litter the licenses metadata with a one-off license just for LD.
  * _Downsides:_ the upstream license isn't FSF blessed and the non-commercial aspect isn't well represented in the package metadata.
* Add a special license just for LDMud (`licenses.ldmud`?)
  * _Upsides:_ fully captures the LDMud specific constraints from the upstream copyright.
  * _Downsides:_ I can't image other packages will use this license, so there's some Nixpkgs-wide bloat.
* Switch it to a non-free package
  * _Upsides:_ resolves the situation without Nixpkgs bloat.
  * _Downsides:_ other than the non-commercial constraint LDMud is what I'd consider free software in the sense we can modify it, redistribute it, etc. It would be a shame to have no binary builds and an extra hurdle for users to jump through
  
What do others think? Is there someone in particular well suited to give guidance here we could tag in?

Thanks!

Reply:
A license is just an attribute set with a pretty name. if you need a custom one [you can do something like what's done in n8n](https://github.com/NixOS/nixpkgs/blob/ac3e43ca8a4f65372657bd9b5c5a28b66921f189/pkgs/applications/networking/n8n/default.nix#L29-L35)

It's not commercially redistributable so it'd just be the same as unfree. (And it doesn't matter much since hydra doesn't build unfree things even if they are redistributable)

I'd just set it to `unfree`

Reply:
I don't have much context on the nixpkgs imperatives here, but at least from reading through https://github.com/NixOS/nixpkgs/issues/83884 it seems like it's probably worth the effort to mark it as redistributable, since it would enable the existing ~unofficial caches to build it for now and would ensure it starts getting built if/when the same applies to nixpkgs itself.

Reply:
if it is redistributable one can set unfreeRedistributable. I'm not sure if it is however.

Reply:
The upstream's [COPYRIGHT](https://github.com/ldmud/ldmud/blob/64b3588a13fb0c761da62d6deb56dfa380b03c6f/COPYRIGHT#L16-L23) file says:

```
Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions
are met:
1. Redistributions of source code must retain the above copyright
   notice, this list of conditions and the following disclaimer.
2. Redistributions in binary form must reproduce the above copyright
   notice, this list of conditions and the following disclaimer in the
   documentation and/or other materials provided with the distribution.
```

I am not sure if 2 is met automagically by the current build, but it seems doable in some form or another. Does ensuring the COPYRIGHT file is copied into the output suffice?

Reply:
Using `unfreeRedistributable` seems reasonable to me as well. :+1: 

I've switched to that license in 65f142b. Separately I also made sure the COPYRIGHT file makes it into the output in 4aad58f.

Reply:
I'm personally not entirely convinced this is unfreeRedistributable.

It says:

```
Permission is granted to use, extend and modify
the source code and any executables created from the source code
provided subject to the restriction that the source code and any
executables created therefrom may not be used in any way whatsoever
for monetary gain.
```

It's not clear to me that this is compatible with the redistribution clause. (For example what if someone earned money from running the cache it was served from? How direct does this have to be?) I'm not a lawyer, so idk what exactly is going on here.

but I'm personally always going to be uncomfortable with licenses which contradict themselves like this.

Reply:
Unfortunately the heredity of this software is quite old (the earliest form was circa ~89) and predates much of the consensus around free software licensing that has evolved since.

>  (For example what if someone earned money from running the cache it was served from? How direct does this have to be?) 

I think it will be hard to address these concerns concretely but my understanding of the context in which the non-commercial clause was added leads me to believe that it's principally concerned with games being run on LDMud with a for-profit model.

> but I'm personally always going to be uncomfortable with licenses which contradict themselves like this.

Judging by the metric of emoji reactions (the most scientific of all metrics) I think @viraptor, @abathur and myself believe we've met the bar for unfreeRedistributable. I'm going to resolve this comment but would be open to more discussion if others disagree strongly.